lunatrace/bsl/ingest-worker/pkg/epss/ingester.go

```diff
@@ -0,0 +1,151 @@
+// Copyright by LunaSec (owned by Refinery Labs, Inc)
+//
+// Licensed under the Business Source License v1.1
+// (the "License"); you may not use this file except in compliance with the
+// License. You may obtain a copy of the License at
+//
+// https://github.com/lunasec-io/lunasec/blob/master/licenses/BSL-LunaTrace.txt
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package epss
+
+import (
+  "context"
+  "database/sql"
+  "fmt"
+  "github.com/rs/zerolog/log"
+  "go.uber.org/fx"
+  "strings"
+)
+
+type EPSSIngester interface {
+  Ingest(ctx context.Context) error
+}
+
+type EPSSIngesterParams struct {
+  fx.In
+
+  DB *sql.DB
+}
+
+type epssIngester struct {
+  deps EPSSIngesterParams
+}
+
+// createEpssScoreLookup Create lookup of EPSS scores by CVE
+func createEpssScoreLookup(epssScores []EpssVulnerability) map[string]EpssVulnerability {
+  lookup := make(map[string]EpssVulnerability)
+  for _, epssScore := range epssScores {
+    lookup[epssScore.Cve] = epssScore
+  }
+  return lookup
+}
+
+// getEpssScore Get the EPSS score for a given CVE
+func getEpssScore(cve string, epssScoreLookup map[string]EpssVulnerability) (EpssVulnerability, error) {
+  epssScore, ok := epssScoreLookup[cve]
+  if !ok {
+    return EpssVulnerability{}, fmt.Errorf("could not find epss score for cve %s", cve)
+  }
+  return epssScore, nil
+}
+
+func (s *epssIngester) Ingest(ctx context.Context) error {
+  // Contains the CVE (id), EPSS (score), and Percentile (ranking) for each vulnerability
+  epssScores, err := FetchEpssScores()
+  if err != nil {
+    log.Error().
+      Err(err)
+    return err
+  }
+
+  epssScoreLookup := createEpssScoreLookup(epssScores)
+
+  tx, err := s.deps.DB.BeginTx(ctx, &sql.TxOptions{
+    Isolation: sql.LevelReadUncommitted,
+  })
+  if err != nil {
+    return err
+  }
+  defer tx.Rollback()
+
+  // Get all CVEs and GHSAs from the database
+  cvePairs, err := tx.QueryContext(ctx, `
+    SELECT v1.source_id, v2.source_id FROM vulnerability.equivalent e
```

Comment:
im glad this works for now, but there must be a real library we can use for this in the future. passing the raw sql strings is starting to look clunky, although god knows ive done it a lot myself

Reply:
In node there are quite a few query builder libraries. There are also libraries which will read in .sql files and template stuff into them. That has the advantage of code separation, and it will allow the IDE to work some autocomplete magic on the SQL. We might consider that in the future

Reply:
![Screenshot_2023-01-03_15-54-53](https://user-images.githubusercontent.com/4573221/210460662-1cd80826-223e-48a9-9d47-f8405484a841.png)

The SQL autocomplete works great for me FWIW